examples/kepler-integration/src/components/export-video/export-video-panel-container.js

```diff
@@ -90,19 +70,35 @@ export class ExportVideoPanelContainer extends Component {
       mediaType: 'GIF',
       cameraPreset: 'None',
       fileName: 'Video Name',
-      //  quality: "High (720p)",
+      quality: 'High (720p)',
       viewState: this.props.mapData.mapState,
+      currentWidth: 1280,
+      currentHeight: 720,
       durationMs: 1000,
       encoderSettings: {
         framerate: 30,
         webm: {
+          // TODO width/height not in WEBMEncoder
           quality: 0.8
         },
         jpeg: {
+          // TODO width/height not in JPEGEncoder
           quality: 0.8
         },
         gif: {
-          sampleInterval: 1000
+          sampleInterval: 1000,
+          get width() {
+            return this.currentWidth;
+          },
+          set width(width) {
+            width = this.currentWidth;
+          },
+          get height() {
+            return this.currentHeight;
+          },
+          set height(height) {
+            height = this.currentHeight;
```

Comment:
While a clever use of scope, this isn't safe to do. People avoid using "this" within plain objects usually. Instead, use a reducer pattern where you make a copy of the encoder settings objects and replace the width and height property (this is what people use "..." for a lot)

Reply:
Could you give me an example? This is what I interpreted but it doesn't give me the width and height within gif 
```javascript
    this.state = {
      currentWidth: 1280,
      currentHeight: 720,
      encoderSettings: {
        gif: {
          sampleInterval: 1000,
          width: {...currentWidth},
          height: {...currentHeight}
        },
```

Reply:
I'll have to look that up. The current method is after a lot of trial and error. Still doesn't give me access to the currentWidth for use in encoderSettings.

Reply:
https://stackoverflow.com/questions/43040721/how-to-update-nested-state-properties-in-react

Reply:
I'm not sure that link fully quite accomplishes what I want to do. I wanted the width and height to be set outside of encoder settings. This way it only needs to be changed once for all formats. I'm imagining the function will account for resolution as well as aspect ratio in the future. A presenter might not want a 16:9 aspect ratio for their slide. Maybe they want a square 1:1 or even 9:16. I believe this implementation would set the groundwork to allow for that.

There's also inconsistency in the current export. For example, GIF will export 16:9 as we set it but WEBM will be 1:1. However, I'm adding that as a different issue.

There's 2 main issues I'm struggling with:
1. Setting a currentHeight/Width state and having the respective format in encoderSettings inherit it. 
2. Having the state be set again and then force updated.

Here it is in code form. This doesn't work but I hope it gets the point across
```javascript
export class ExportVideoPanelContainer extends Component {
  constructor(props) {
    this.state = {
      currentWidth: 1280, // this should be updated as user picks settings
      currentHeight: 720, // this should be updated as user picks settings
      encoderSettings: {
        gif: {
          sampleInterval: 1000,
          width: this.currentWidth,
          height: this.currentHeight,
        },
        filename: 'kepler.gl'
      },
      adapter: new DeckAdapter(this.getDeckScene)
    };
  }

setQuality(resolution) {
    this.setState({
      quality: resolution
    });
    if (resolution === 'Good (540p)') { // Change resolution once so all formats can use it
      this.currentWidth = 960
      this.currentHeight = 540
      this.forceUpdate()
  }
```

Reply:
Oh this is an anti-pattern. That's why React has been fighting me so hard on this. https://stackoverflow.com/a/49529378 . Didn't know that constructor was only called once. I thought forceUpdate would update the constructor as well

Reply:
oh I think this is an anti pattern https://stackoverflow.com/a/49529378. I'll have to think of another way to do this then...

Reply:
Define with simple constructions: 
```
this.state = {
      mediaType: 'GIF',
      cameraPreset: 'None',
      fileName: 'Video Name',
      quality: 'High (720p)',
      viewState: this.props.mapData.mapState,
      currentWidth: 1280,
      currentHeight: 720,
      durationMs: 1000,
      encoderSettings: {
        framerate: 30,
        webm: {
          // TODO width/height not in WEBMEncoder
          quality: 0.8
        },
        jpeg: {
          // TODO width/height not in JPEGEncoder
          quality: 0.8
        },
        gif: {
          sampleInterval: 1000,
          width: 1280,
          height: 720
        },
        filename: 'kepler.gl'
      },
      adapter: new DeckAdapter(this.getDeckScene)
    };
```

Update like this:
```
let newWidth = 500
let newHeight = 300
this.setState({
  currentWidth: newWidth, 
  currentHeight: newHeight, 
  encoderSettings: {
    ...this.state.encoderSettings,
    gif: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    }
})

Reply:
I was hoping to avoid that way of updating as that'd get very long in the future if more formats are supported. I wanted all of it to be referenced within state so that I could just setState on currentHeight/Width once to have all formats updated. However, I'll accept defeat for today as I've spent a lot of time researching and it doesn't seem to be possible to do.

For example:
```javascript
// Multiply this amount of code by however many resolutions are supported worst case.
let newWidth = 500
let newHeight = 300
this.setState({
  currentWidth: newWidth, 
  currentHeight: newHeight, 
  encoderSettings: {
    ...this.state.encoderSettings,
    gif: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    },
    jpeg: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    },
    webm: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    },
    png: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    },
    mp4: {
        ...this.state.encoderSettings.gif,
        width: newWidth,
        height: newHeight
    },
})
```

Reply:
I can relate to having that reaction too, but ultimately I've found it never gets too bad because people will start splitting long updates into functions. The punch line is these functions are called reducers :D